src/script/interpreter.cpp

```diff
@@ -335,39 +335,109 @@ bool EvalScript(vector<vector<unsigned char> > &stack,
-    CScript::const_iterator pend = script.end();
-    CScript::const_iterator pbegincodehash = script.begin();
-    opcodetype opcode;
-    valtype vchPushValue;
-    vector<bool> vfExec;
-    vector<valtype> altstack;
+    ScriptMachine sm(flags, checker);
+    sm.setStack(stack);
+    bool result = sm.Eval(script);
+    stack = sm.getStack();
+    if (serror)
+        *serror = sm.getError();
     if (sighashtype)
-        *sighashtype = 0;
+        *sighashtype = sm.getSigHashType();
+    return result;
+}
 
-    set_error(serror, SCRIPT_ERR_UNKNOWN_ERROR);
-    if (script.size() > MAX_SCRIPT_SIZE)
-        return set_error(serror, SCRIPT_ERR_SCRIPT_SIZE);
-    int nOpCount = 0;
+static const CScriptNum bnZero(0);
+static const CScriptNum bnOne(1);
+static const CScriptNum bnFalse(0);
+static const CScriptNum bnTrue(1);
+static const StackDataType vchFalse(0);
+static const StackDataType vchZero(0);
+static const StackDataType vchTrue(1, 1);
+
+// Returns info about the next instruction to be run
+std::tuple<bool, opcodetype, StackDataType, ScriptError> ScriptMachine::Peek()
+{
+    ScriptError err;
+    opcodetype opcode;
+    StackDataType vchPushValue;
+    auto oldpc = pc;
+    if (!script->GetOp(pc, opcode, vchPushValue))
+        set_error(&err, SCRIPT_ERR_BAD_OPCODE);
+    else if (vchPushValue.size() > MAX_SCRIPT_ELEMENT_SIZE)
+        set_error(&err, SCRIPT_ERR_PUSH_SIZE);
+    pc = oldpc;
+    bool fExec = !count(vfExec.begin(), vfExec.end(), false);
+    return std::tuple<bool, opcodetype, StackDataType, ScriptError>(fExec, opcode, vchPushValue, err);
+}
+
+
+bool ScriptMachine::BeginStep(const CScript &_script)
+{
+    script = &_script;
+
+    pc = pbegin = script->begin();
+    pend = script->end();
+    pbegincodehash = pc;
+
+    sighashtype = 0;
+    nOpCount = 0;
+    vfExec.clear();
+
+    set_error(&error, SCRIPT_ERR_UNKNOWN_ERROR);
+    if (script->size() > MAX_SCRIPT_SIZE)
+    {
+        script = nullptr;
+        return set_error(&error, SCRIPT_ERR_SCRIPT_SIZE);
+    }
+    return true;
+}
+
+
+int ScriptMachine::getPos() { return (pc - pbegin); }
+bool ScriptMachine::Eval(const CScript &_script)
+{
+    bool ret;
+
+    if (!(ret = BeginStep(_script)))
+        return ret;
+
+    while (pc < pend)
+    {
+        ret = Step();
+        if (!ret)
+            break;
+    }
+    if (ret)
+        ret = EndStep();
+    script = nullptr; // Ensure that the ScriptMachine does not hold script for longer than this scope
+
+    return ret;
+}
+
+bool ScriptMachine::EndStep()
+{
+    script = nullptr; // let go of our use of the script
+    if (!vfExec.empty())
+        return set_error(&error, SCRIPT_ERR_UNBALANCED_CONDITIONAL);
+    return set_success(&error);
+}
+
+bool ScriptMachine::Step()
+{
     bool fRequireMinimal = (flags & SCRIPT_VERIFY_MINIMALDATA) != 0;
+    opcodetype opcode;
+    StackDataType vchPushValue;
+    ScriptError *serror = &error;
     try
     {
-        while (pc < pend)
+        if (1)
```

Comment:
since `pc < end` is enforced by the caller (`ScriptMachine::Eval`) why not removing the `if (1)` and the block defined by the now useless (?) curly brackets?

Reply:
I knew you were going to mention this :-).  The reason why is without the if (1), the entire interpreter logic will get indented differently.  I believe that this would be bad both for this PR and long term.  For this PR, the entire interpreter will look changed.  For long term, it'll be harder for us to merge new OP code implementations coming from other clients.

Reply:
ok so it was not a "functional" reason, I thought I could have miss something. 

That said why not just removing the `if (1)` and leave the brackets as is? The indentation should be preserved. 

In any case I would add a comment to explain the reason why the `if(1)` has been placed there. 

Lastly speaking of porting code from other implementation IMHO one of the main things that made it really difficult is the way we format curly brackets to define block of code in `if` branches, loop etc etc. The majority of other projects use `BreakBeforeBraces: Attach
` whereas we use `BreakBeforeBraces: Allman`.

Reply:
that's a good point.  The if(1) is I guess an old school technique that used to be used because people did not like the look of naked scopes.  However, with the greater use of ctor/dtors and scoped lifecycles people are much more used to naked scopes.  I will fix.